### PR TITLE
http filter manager: handle trailers added during continuation

### PR DESCRIPTION
**Pull request by mattklein123**
The following sequence lead to trailers being continued twice:
1) Filter 1 stops iteration in headers
2) Filter 1 continues iteration on data with end stream true
3) Filter 2 continues iteration in headers
4) filter 2 continued iteration in data, but adds trailers

Previously trailers were continued twice, now they are continued
once.

Risk Level: Medium
Testing: New UT
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: source/common/http/filter_manager.cc
Comment (alyssawilk) [inline source/common/http/filter_manager.cc:104]: lazy ask, can we have a similar problem adding trailers while processing headers / metadata or is it only a data problem?

**Review by alyssawilk — APPROVED** [review]
thanks for debugging and fixing!  LGTM modulo the question I had on the other PR.

File: source/common/http/filter_manager.cc
Comment (mattklein123) [reply source/common/http/filter_manager.cc:104]: I'm traveling today/tomorrow and in all day meetings. I'm inclined to get this fix in prior to release and I can research your question separately? Or do you want to hold this until I investigate? I think this fix will cover the header case also, but I need to do some research.

Comment (alyssawilk) [reply source/common/http/filter_manager.cc:104]: SGTM.  Bonus points for TODO or tracking bug but really I trust you're good for it :-P

Comment (alyssawilk) [reply source/common/http/filter_manager.cc:104]: Also if we're adding close to the release maybe runtime guard just in case there's bugs?  Optional, because I'm happy to defer to your judgement here.

**Review by snowp — APPROVED** [review]
LGTM, thanks!

**Review by yanavlasov — APPROVED** [review]

File: source/common/http/filter_manager.cc
Comment (mattklein123) [reply source/common/http/filter_manager.cc:104]: I looked into this and currently we only "allow" Add*Trailers() to be called during decodeData()/encodeData() when end_stream is true. This is tracked by the following assertion:

```
// Trailers can only be added during the last data frame (i.e. end_stream = true).
ASSERT(state_.filter_call_state_ & FilterCallState::LastDataFrame);
```

Given that we don't support adding trailers in headers/metadata today, I think it's probably safe to move forward with this fix. I'm OK without the runtime guard given how targeted the fix is and that without it the code will crash or do something else very bad so I'm going to go ahead and merge.

**Pull request merged**
